src/Lecture2.hs

```diff
@@ -288,8 +397,48 @@ It also can be:
 
 x + 45 + y
 
-Write a function that takes and expression and performs "Constant
+Write a function that takes an expression and performs "Constant
 Folding" optimization on the given expression.
 -}
+{- Well, this solution gets the test to succeed, but that's just it.
+   An expression with more Var components would still go into infinite mode.
+   So there has to be a better way ... 
+   Any hints?? :-) -}
```

Comment:
One possible solution is to reduce `Expr` to a data type like `([String], Int)` -- a pair of all variable names and sum of all constants. And then, you can pattern match on this result and handle different cases (no variables, the sum is zero) to produce the result.

You can traverse the `Expr` data type and collect all the results in such an accumulator. This will also use more precise data types instead of just `Expr`.

For example, `varsOnly` function has the following type:

```haskell
varsOnly :: [Expr] -> [Expr]
```

However, it's not precise. `varsOnly` tells that it returns list of `Expr` however it can't return neither `Lit` nor `Add` constructors. The function filters by some criteria but then discards this knowledge. A more precise type would be:

```haskell
varsOnly :: [Expr] -> [String]
```

But, in a solution with `([String], Int)` you don't need `varsOnly` function.

There's an alternative solution without intermediate data structure that relies only rebalancing trees but I find it a bit harder to implement 🙂